src/include/daos_fs_sys.h

```diff
@@ -0,0 +1,454 @@
+ * \param[in]	obj	Object to release.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_close(dfs_obj_t *obj);
+
+/**
+ * Read data from the file object, and return actual data read.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	obj	Opened file object.
+ * \param[in,out]
+ *		buf	[in]: Allocated buffer for data.
+ *			[out]: Actual data read.
+ * \param[in]	off	Offset into the file to read from.
+ * \param[in,out]
+ *		size	[in]: Size of buffer passed in. [out]: Actual size of
+ *			      data read. On error, this is set to -1.
+ * \param[in]	ev	Completion event, it is optional and can be NULL.
+ *			Function will run in blocking mode if \a ev is NULL.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_read(dfs_sys_t *dfs_sys, dfs_obj_t *obj, void *buf, daos_off_t off,
+	     daos_size_t *size, daos_event_t *ev);
+
+/**
+ * Write data to the file object.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	obj	Opened file object.
+ * \param[in]	buf	Data to write.
+ * \param[in]	off	Offset into the file to write to.
+ * \param[in,out]
+ *		size	[in]: Size of buffer passed in. [out]: Actual size of
+ *			      data written. On error, this is set to -1.
+ * \param[in]	ev	Completion event, it is optional and can be NULL.
+ *			Function will run in blocking mode if \a ev is NULL.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_write(dfs_sys_t *dfs_sys, dfs_obj_t *obj, const void *buf,
+	      daos_off_t off, daos_size_t *size, daos_event_t *ev);
+
+/**
+ * Punch a hole in the file starting at offset to len. If len is set to
+ * DFS_MAX_FSIZE, this will be a truncate operation to punch all bytes in the
+ * file above offset. If the file size is smaller than offset, the file is
+ * extended to offset and len is ignored.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	file	Link path of file.
+ * \param[in]	offset	offset of file to punch at.
+ * \param[in]	len	number of bytes to punch.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_punch(dfs_sys_t *dfs_sys, const char *file,
+	      daos_off_t offset, daos_off_t len);
+
+/**
+ * Similar to dfs_sys_punch but on an open object.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	obj	Opened file object.
+ * \param[in]	offset	offset of file to punch at.
+ * \param[in]	len	number of bytes to punch.
+ *
+ * \return		0 on success, errno code on failure.
+ */
+int
+dfs_sys_opunch(dfs_sys_t *dfs_sys, dfs_obj_t *obj,
+	       daos_off_t offset, daos_off_t len);
+
+/**
+ * Remove an object identified by path. If object is a directory and is
+ * non-empty; this will fail unless force option is true. If object is a
+ * symlink, the symlink is removed.
+ *
+ * \param[in]	dfs_sys Pointer to the mounted file system.
+ * \param[in]	path	Link path of object.
+ * \param[in]	force	If true, remove dir even if non-empty.
```

Comment:
What does force do for non-empty directories?  This isn't the same as recursive remove I assume, it simply abandons the entries?  That may not be desirable behaviour to expose.

Reply:
For non-empty directories, `force = true` _**will**_ recursively remove all entries. This option is just passed along to `dfs_remove()`:
https://github.com/daos-stack/daos/blob/1f1b84698223a3085fb643700d901a40277cd409/src/include/daos_fs.h#L505-L520
https://github.com/daos-stack/daos/blob/f0d81c841ba27f21eaf8b4eaf8225d1a00f94c85/src/client/dfs/dfs.c#L1936-L1940

Do you think:
1. `dfs_remove()` should not do this either?
2. `dfs_sys_remove()` should just not expose `force`?

Reply:
A comment about the effect of the arg, even if it just says "consult dfs_remove() for details".